### PR TITLE
workspace/files and textDocument/content extensions

### PR DESCRIPTION
**Pull request by felixfbecker**
@sqs what do you think of this?
Basically it is a simple extension to serve the minimal use case of globbing the workspace for indexing and getting the content of files.
In opposite to #3 it does not introduce the concept of a "virtual file system" which is nice because it allows an IDE to operate solely on another protocol like FTP.

I only went by my use case - are there features missing that other LS need?

ToDo
- [x] extend `ClientCapabilites`
- Allow binary files to be retrieved with Base64 encoding (doesn't really match with "textDocument")?
- More [glob options](https://www.npmjs.com/package/glob#options)?
- enforce support for glob features like `**`?

### CODE REVIEW

File: protocol.md [hunk 1]
```diff
@@ -837,6 +837,71 @@ interface TextDocumentContentChangeEvent {
...
+```
+
+_Response_:
+* result: `string`
+* error: code and message set in case an exception happens during showing a message.
+
+#### Glob Request
+
+The glob request is sent from the server to the client to request a list of files in the workspace that match a glob pattern.
+The glob pattern must be matched only against the path of the files in the workspace and relative to the `rootPath`.
```
Comment (sqs) [inline protocol.md]: Lots of LS need to fetch files outside the `rootPath`. For example, the rootPath might be file:///gopath/src/github.com/foo/bar but Go needs to fetch from the GOROOT (/usr/local/go/src), Python might need to fetch from /usr/lib/python/whatever, etc. You could specify a pattern of `../../../**/*.php` but it's also not clear if those files are "in the workspace."

**Comment by sqs** [comment]
I like this. I think this is better, and I can't think of a reason why it wouldn't work.

> I only went by my use case - are there features missing that other LS need?
- There are some edge cases where `go` cares about whether something is a symlink. I suppose this is the case for some other tools. Windows doesn't really support symlinks, and most things must support Windows, so it might be a non-issue.
- As mentioned, it needs to be able to access anything in `file:///` potentially, not just underneath the workspace rootPath. Minor thing, very fixable.
- For Go, this will be slightly more complex to implement than the current VFS extension, and about the same as #3. The current VFS is never going to be fast enough anyway, so it's not an acceptable status quo anyway, just throwing it out for comparison.

Other good things:
- The extra RTT introduced before you can read files (first you must call glob to get the actual filenames; you can't pass a glob and get file contents back) is not a performance problem at all. And it simplifies the API because you don't need to return a map whose keys are the URIs.  :+1: 

Open questions:
- If you need to fetch a lot of file contents, this will have many of the same perf problems as the current VFS because you'll have a lot of serial operations. Can you make the `content` method accept a list of URIs? That is kind of reimplementing JSON-RPC 2.0 batching at the application layer, but vscode doesn't implement batching, so we would need to do this (I think).

@keegancsmith @slimsag  WDYT?

**Comment by felixfbecker** [comment]
> Windows doesn't really support symlinks

I use symlinks every day on Windows to replace dependencies with local development versions, so I think this is important on every OS. `mklink` (cmd) and `New-Symlink` (Powershell/PSCX) FTW

It might be worth to support more options for glob. Some inspiration:
https://www.npmjs.com/package/glob#options

For example, `dot` might be interesting.

**Comment by felixfbecker** [comment]
> For Go, this will be slightly more complex to implement than the current VFS extension, and about the same as #3

In what way?
Also wanted to point out that a glob request like `directory/*.*` would basically be the equivalent to `vfs/readDir`.

**Comment by sqs** [comment]
Yep. It'd be slightly more complicated because the current fs/\* ops were literally designed around what Go needs. To make this work, it'd need to somehow also get dirs (glob only will return files here, right?) because Go does call stat on dirs and expects subdirs to be returned by readDir. Little things like that - nothing fundamentally hard.

**Comment by felixfbecker** [comment]
Found this in the node glob docs:

> `nodir` Do not match directories, only files. (Note: to match only directories, simply put a / at the end of the pattern.)

Also, regarding symlinks:

> `follow` Follow symlinked directories when expanding *\* patterns. Note that this can result in a lot of duplicate references in the presence of cyclic links.

**Comment by felixfbecker** [comment]
A simple test to only get directories:

```
$ glob */
bin
fixtures
images
src
tests
vendor
```

`glob *` will return both directories and files. We should require a trailing slash for directories to differentiate, and maybe support the `nodir` option.

**Comment by felixfbecker** [comment]
> If you need to fetch a lot of file contents, this will have many of the same perf problems as the current VFS because you'll have a lot of serial operations. Can you make the content method accept a list of URIs? That is kind of reimplementing JSON-RPC 2.0 batching at the application layer, but vscode doesn't implement batching, so we would need to do this (I think).

Well in my case I will request the list of files, get file 1, parse file 1, get file 2, parse file 2... Because parsing takes quite some time this is actually more performant than prefetching. Afaik the LSP doesn't state anywhere that Batch is not supported, this is only not used by VS Code... How about I make sure it is supported by PHP LS and you just use it? All other textDocument/\* methods apply to a single text document

**Comment by alexsaveliev** [comment]
@felixfbecker, @sqs  I have few comments here:
- you are adding new methods into the existing namespaces, might be a problem when proposed to protocol's owner :)
- @sqs I don't feel confident enough allowing access to files outside of the `rootPath`. We can't restrict people working on server implementations from making mistakes and exposing sensitive file contents. I'd prefer to use some magic URI (for example `gosrc://io/io.go`). 
- I'd prefer to extend a bit `GlobParams` to have ability to control `dot` and `ignore` glob options (have no plans to use them now, however). For example, we probably should ignore composer's `vendor` directory when searching for PHP files.
- For my needs (sorry guys :) ) I'd like to have functionality proposed by @sqs (`Can you make the content method accept a list of URIs?`) and ability to specify multiple patterns

**Comment by felixfbecker** [comment]
@alexsaveliev 

> you are adding new methods into the existing namespaces, might be a problem when proposed to protocol's owner :)

Well that was the idea, add these methods to existing namespaces where they make sense. But `workspace` is actually not used atm for the client.
Before this makes it into the standard I would implement them as `workspace/_glob` for example to avoid conflicts with the final standard.

> I don't feel confident enough allowing access to files outside of the rootPath. We can't restrict people working on server implementations from making mistakes and exposing sensitive file contents. I'd prefer to use some magic URI (for example gosrc://io/io.go). 

How would that URI work with an IDE like Eclipse or VS Code work though? I think many LS will have the need to read files outside the workspace, like config files in the users home dir.

> I'd prefer to extend a bit GlobParams to have ability to control dot and ignore glob options (have no plans to use them now, however).

I will add an `options` parameter that includes a subset of the node-glob options: `dot` and `nodir`. Anything else needed?

> For example, we probably should ignore composer's vendor directory when searching for PHP files.

vendor must actually be indexed, otherwise I cannot give you definition of library classes that are used.

> For my needs (sorry guys :) ) I'd like to have functionality proposed by @sqs (Can you make the content method accept a list of URIs?) and ability to specify multiple patterns

Multiple patterns is a great idea, this will also solve ignore, because you can prefix globs with `!` to negate them.

For getting many contents at once, would JSON RPC Batch work for you?

**Comment by renfredxh** [comment]
> For getting many contents at once, would JSON RPC Batch work for you?

@felixfbecker possibly. Looking into it now

**Comment by alexsaveliev** [comment]
@felixfbecker 

> I will add an options parameter that includes a subset of the node-glob options: dot and nodir. Anything else needed?

should be enough for the start, I think

> For getting many contents at once, would JSON RPC Batch work for you?

I'd rather use some method that accepts multiple input parameters. My reasons are the following:
- at least for java and js/ts we'll have to modify two repositories (lsp server; language support) to add batch JSON RPC while we can reduce amount of work needed if we'll just add yet another method(s). We may face the same with other languages as well
- JSON RPC spec states that batch requests may be processed in any order. It may add a couple of potential issues if we'll introduce batch request support if it wasn't in the library used.

Otherwise LGTM

**Comment by felixfbecker** [comment]
My problem with multiple URIs at once is that all other `textDocument/*` methods get a `TextDocumentIdentifier` as a parameter. Basically you call the method on a specific text document, like `someTextDocument.content()`. It wouldn't make much sense to call the content method on an array of text documents, this would not be in line with how the `textDocument` namespace is used in the rest of the protocol. You would have to return the results in an array then too. But what about an error case? The `error` property cannot be an array, you would have to use `data` to report which operation failed, and it doesn't really make sense to consider the whole request failed if just one of the contents could not be received. `data` however is not standardized, so you would have to spec that in the LSP first. At which point you are basically re-specifying Batch. Btw, you could also just send multiple requests in parallel, your JSON RPC implementation should be able map the response ID back to the request 🙂

**Comment by sqs** [comment]
So, basically this proposal would change the current `fs/{readFile,readDir,stat,lstat}` as follows:
1. Rename `fs/readFile` to `textDocument/content` and standardize the params with the rest of LSP
2. Remove `fs/{readDir,stat,lstat}` and replace them with `workspace/glob`, which makes it possible to turn an arbitrary number of `fs/*` calls into a single `workspace/glob` call (potential for huge perf improvement)

@felixfbecker To simplify the `workspace/glob` call parameters, can we make it:
- always include dirs; remove the `noDir` param option
- always include dot files (it is easy to filter those out on the lang server side if they are unneeded); remove the `noDir` param option
- instead of returning a `TextDocumentIdentifier[]`, return a `FileIdentifier[]` with:
  `typescript
  interface FileIdentifier extends TextDocumentIdentifier {
  type: "file" | "dir" | "symlink"; // in reality, this would be an enum, but they're strings here for simplicity
  }
  `

That seems to work for all known LS implementations to dramatically improve perf. It's simple and easy to implement.

**Comment by keegancsmith** [comment]
As part of this proposal we should explicitly document what glob patterns we support, since what a glob does varies across languages / shells. I agree with the simplifications mentioned, but it still leaves open things like `**`, or things like BASH's `extglob`. Instead of saying what it is not, we should explicitly document what it is.

**Comment by felixfbecker** [comment]
@sqs 

> So, basically this proposal would change the current `fs/{readFile,readDir,stat,lstat}` as follows:
> 1. Rename `fs/readFile` to `textDocument/content` and standardize the params with the rest of LSP
> 2. Remove `fs/{readDir,stat,lstat}` and replace them with `workspace/glob`, which makes it possible to turn an arbitrary number of `fs/*` calls into a single `workspace/glob` call (potential for huge perf improvement)

Correct. The motivation is to decouple the API from the notion of a file system, since the protocol works with URIs. So we must think about _resources_ and _collections_ instead of files and folders. We can get the content of a text document resource and apply a glob pattern to multiple resources/collections.

> always include dirs; remove the noDir param option

We could do that if it complicates the client implementation. It would result in a bit more overhead if the server is not interested in directories and additional filtering. I assume the go-to Go glob implementation does not have such an option? I just know about node-glob (which supports this, and is probably used by VS Code)

> always include dot files (it is easy to filter those out on the lang server side if they are unneeded); 

Yes, could do that as well with the same reasoning. Can also be easily accomplished with adding `!**/.*` to the pattern array.

> instead of returning a TextDocumentIdentifier[], return a FileIdentifier[] with: typescript interface FileIdentifier extends TextDocumentIdentifier { type: "file" | "dir" | "symlink"; // in reality, this would be an enum, but they're strings here for simplicity }  

Not sure about this one. This would introduce the concept of files and directories again. A server can easily see if the item is a directory (collection) by checking if the URI path ends with a slash (see example). For symlinks, at least node-glob doesn't even give you that info. Afaik a glob implementation is basically recursive `readdir` operations, which gives you only the name of the entries. To find out if it is a symlink, the client would have to stat every single result. That is quite a performance hit if you are globbing 10,000 files and the server is not interested in this info. Could you explain the use case for this a bit?

@keegancsmith 

> As part of this proposal we should explicitly document what glob patterns we support, since what a glob does varies across languages / shells. I agree with the simplifications mentioned, but it still leaves open things like **, or things like BASH's extglob. Instead of saying what it is not, we should explicitly document what it is.

Yes, I agree. I am just not so familiar with the different "standards" out there. But the client must support globstar imo because that is the biggest use case, and a glob request without a `**` must only match one directory level. I don't have a specific use case for any other extension, so I would consider these optional. For example, if I let the user (client) configure some includes/excludes with globs, it would be nice to support as many as possible, but that is the client's decision.

**Comment by felixfbecker** [comment]
I implemented the current proposal in the PHP LS with an `x` prefix
https://github.com/felixfbecker/php-language-server/pull/136

**Comment by sqs** [comment]
@felixfbecker Great. We will implement the client side of the current proposal soon (likely today), and the Go/TypeScript/JavaScript implementations as well.

> (Re: including glob result "type" (dir/file/symlink) Not sure about this one. This would introduce the concept of files and directories again. A server can easily see if the item is a directory (collection) by checking if the URI path ends with a slash (see example). For symlinks, at least node-glob doesn't even give you that info. Afaik a glob implementation is basically recursive readdir operations, which gives you only the name of the entries. To find out if it is a symlink, the client would have to stat every single result. That is quite a performance hit if you are globbing 10,000 files and the server is not interested in this info. Could you explain the use case for this a bit?

I see your point, and these fields would only be necessary to handle relatively rare edge cases (I think).  Let's move forward without including the `type` field and see if there are unforeseen negative consequences.

**Comment by felixfbecker** [comment]
@sqs What do you think about the `GlobOptions`? Would the implementation of the two options complicate things a lot?

**Comment by sqs** [comment]
@felixfbecker:

Do we need glob negation? I notice your PR doesn't use it. I don't see it being necessary, given that:
1. Most glob implementations probably don't optimize to skip traversals if they would be negated, so it doesn't save time on the client.
2. The common case for LSes is something like "get me all files with extensions .foo or .bar". In this case, you'd prefer for GlobParams to be `string[]` with globs that are OR'd, not AND'd. If we allow negations, we must make it AND'd, and this would require multiple roundtrips for this common case (or supporting `{foo,bar}`). (I think it's rare that LSes would want to negate `!vendor/**`...the Go/PHP/JS/TS LSes do not negate that.)

**Comment by sqs** [comment]
@felixfbecker The GlobOptions would increase the number of combinations we'd need to test/support. It also seems that we could choose one standard behavior (instead of having GlobOptions and making it configurable) with no loss of expressiveness/performance. That's my concern with it.

**Comment by felixfbecker** [comment]
Well removing `nodir` _would_ result in additional overhead for me, because I am not interested in directories. For node-based clients they could just pass the options straight to node-glob, I don't know how easy that is for whatever glob impl is popular in Go. Should I remove the options argument?

Another thing, any idea for a different name for the glob result items? `TextDocumentIdentifier` is confusing because it doesn't have to be a text document.

**Comment by sqs** [comment]
@felixfbecker What about just omitting directories altogether in every case? When Sourcegraph runs LSes, the file system is backed by a git repo, so directories would be inferred anyway. LSes can still, of course, hit the file system if they need to have access to full POSIX, but this seems like a good compromise for the sake of this extension.

In that case, the glob results are basically text documents, so `TextDocumentIdentifier` is OK. (It could also match, e.g., `.jar` files, but that's another issue and I think we can pretend they are text documents for now for the sake of this extension.)

I have an updated draft spec that I will push shortly.

**Comment by felixfbecker** [comment]
> What about just omitting directories altogether?

Can't think of one right now but I have the feeling there are LS out there with a use case for this? Especially since the initial proposal included a `readdir` operation. I think it's good to stay close to standard glob implementations so devs don't get confused, and the default behavior afaik is to include directories (with trailing slash) and exclude dot files, hence the optional `GlobOptions` to configure that.

**Comment by sqs** [comment]
@felixfbecker OK, check the commits I just pushed. You'll need to change the `pattern` field to `patterns`, but otherwise your PHP LS should 100% work. Do you see any other issues?

I looked into the readdir-related system calls, and it seems that on Linux (ext4, btrfs) and BSD (including macOS), it does not add any additional work to determine if a dir entry is a file/dir/symlink. The node-glob library throws away this info, though, and it must call stat again. Sourcegraph will use Go and vscode (if/when it supports this) has its own glob implementation, so node-glob's inefficiency is not a concern for any current real-world, perf-sensitive glob implementations. If/when it becomes a concern, we can patch node-glob. Sound good?

**Comment by felixfbecker** [comment]
@sqs Re: negations
Just saw this note: https://github.com/isaacs/node-glob/blob/master/README.md#comments-and-negation

> Comments and Negation
> Previously, this module let you mark a pattern as a "comment" if it started with a # character, or a "negated" pattern if it started with a ! character.
> These options were deprecated in version 5, and removed in version 6.
> To specify things that should not match, use the ignore option.

So I'm +1 for not supporting negation, but an `ignore` option would still be nice. Actually though now I doubt if we even need multiple patterns? Because there is already glob syntax to handle various combinations of glob patterns:

> - `!(pattern|pattern|pattern)` Matches anything that does not match
>   any of the patterns provided.
> - `?(pattern|pattern|pattern)` Matches zero or one occurrence of the
>   patterns provided.
> - `+(pattern|pattern|pattern)` Matches one or more occurrences of the
>   patterns provided.
> - `*(a|b|c)` Matches zero or more occurrences of the patterns provided
> - `@(pattern|pat*|pat?erN)` Matches exactly one of the patterns
>   provided

I just bring up node-glob because it's the only glob implementation I know well and I thought it would make a nice reference implementation.

Reviewing your changes right now.

File: extension-glob-content.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ClientCapabilities {
+	/* ... all fields from the base ClientCapabilities ... */
+
+	/**
+	 * The client provides support for workspace/xglob.
```
Comment (felixfbecker) [inline extension-glob-content.md]: `x` prefix is used inconsistently in this new spec - I would propose to spec it under the non-prefixed name (and propose it to Microsoft as such in a PR) but use the prefix in the actual implementation until it is officially part of the protocol.

**Comment by sqs** [comment]
@felixfbecker We have to write our own globbing lib in Go (or reuse an existing one, none of which support this advanced node-glob syntax), vscode uses its own globbing lib at https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/src/vs/base/common/glob.ts, and I don't see any node-glob-equivalent libs for Java (which Codenvy/RedHat might need to write), so I want to keep the allowable syntax simple initially.

File: extension-glob-content.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+```
+
+### Glob Request
+
+The glob request is sent from the server to the client to request a list of files in the workspace that match a glob pattern. The glob pattern must be matched only against the path (not other fragments of the URI) and must be relative to the workspace's `rootPath` (set during initialization).
+
+The glob patterns must be interpreted according to the following rules:
+* `*` matches any sequence of non-path separator characters
+* `**` when alone in a path component, matches all files and zero or more directories and subdirectories ("globstar"); does not walk symlinks
+* `![]{}?` characters are interpreted literally and do not have any special meaning (TEMP NOTE: this is to simplify the implementation of globbing; we will need to write our own simple Go globbing library, and a brief survey of many languages' globbing libraries shows wide disparities in behavior)
```
Comment (felixfbecker) [inline extension-glob-content.md]: Can we change this so that a client _may_ give these special meaning? Otherwise it will be a breaking change to add this feature in the future

**Comment by sqs** [comment]
@felixfbecker made those two changes you just commented about :+1:

**Comment by felixfbecker** [comment]
> We have to write our own globbing lib in Go (or reuse an existing one, none of which support this advanced node-glob syntax), vscode uses its own globbing lib at https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/src/vs/base/common/glob.ts, and I don't see any node-glob-equivalent libs for Java (which Codenvy/RedHat might need to write), so I want to keep the allowable syntax simple initially. 

Yes, agree. Implementation should be as easy as possible to make adoption easier. What exactly are you referring to with "advanced node-glob syntax"? The globstar is widely supported afaik

**Comment by sqs** [comment]
> "advanced node-glob syntax"

I just mean the `?(...|...)`, `@(...|...)`, etc.

**Comment by sqs** [comment]
# Open questions

#### Do we need `workspace/glob` to return directories, or can it be only files?

I am leaning toward "only files."

Returning dirs is only necessary (and not derivable from a list of files) if an empty directory is meaningful (it is for Go, but it couldn't be checked into a Git repo, so it's of minimal value) or if it would improve perf a lot to support listing dirs without files (hard to imagine this making a difference).

The downside is that it adds complexity to the spec and the implementation (adding trailing slashes to dir paths, or adding a new `type: "dir" | "file"` field). These can come in a future LSP extension if they become necessary.

#### Not globbing at all?

Crazy idea: can we just make `workspace/glob` (perhaps renamed) just return all filenames, and add globbing in a future extension? Lang servers can just post-filter the list of filenames for now. This adds almost no overhead in the common case.

#### Explicit specification of glob capabilities

Currently the spec says that `!?[]{}` are undefined but will be defined before the spec is proposed upstream.

Anything we support must have a consistent implementation across many languages/platforms:
- Node.js ([node-glob](https://github.com/isaacs/node-glob) is quite advanced already, so we'd probably need to disable some of its features)
- vscode (uses [its own glob lib](https://sourcegraph.com/github.com/Microsoft/vscode/-/blob/src/vs/base/common/glob.ts))
- Go (has some advanced glob libs that we could reuse)
- Java (Eclipse and Eclipse Che probably will want to implement this extension)

So, the simpler the better.

It is not especially valuable to have complex glob syntax. It is easy to post-filter a list of filenames with your exact criteria in your language server, and there's rarely any harm in getting too many filenames (since you typically don't treat all glob results as entrypoints).

Can anyone justify why they would need:
- `!` (negation): could be used to, e.g., omit `node_modules` or `vendor` (does this really yield a significant speedup in practice? Prefetching these might add a few hundred msec but it saves you from many almost certain roundtrips. Can we address this in a future extension?)
- `?` (matches any single non-`/` character)
- `[c-d]` or `[^c-d]`
- `{a,bc}`: nice to have for `**/*.{js,jsx}` but not necessary

**Comment by felixfbecker** [comment]
Hm, hard to judge now for possible later use cases. We could of course introduce the options parameter later and add the inverse of `nodir` - `dir` - to include directories. But we should also think about what devs expect from a `workspace/glob` request, it will be confusing if it behaves very differently from the any other glob implementation.

We could of course just do a `workspace/listFiles` request. However that would be very limited for further extensions in the future. I have open feature requests for both _ignoring_ certain files/folders from indexing, but also for _including_ dependencies outside the root path. I would like to support these by providing settings to the user where they can define glob patterns (VS Code itself also makes heavy use of globs in their settings for configuring excludes etc). I would then just append the include pattern to the `**/*.php` pattern array and add the ignore through a possible future `ignore` option (after seeing the deprecation notice I don't think it is a good idea to negate with `!`, the pattern array semantic should be OR).
Performance does matter in this case. Including or excluding `vendor` for example is a difference of 80 files vs 2000 files in the PHP LS - and that is a small project compared to others. For magento, it is the difference of 17,000 vs 24,000 files. I can filter these myself of course, but such a difference matters even on the protocol level (overhead) and client (if the client implementation has support for not traversing `vendor` at all). This request will be done on every initialization.

Regarding additional pattern syntax, as said I would like to accept globs from the user. So I think we should require the minimal syntax required for the LS itself (globstar, and dot/dir cases needs to be clear) and then let the client support additional syntax if it's available. That means a client can just pass it to a library of his choice and without having to strip anything out. We could make common special chars "reserved" so we can add this in the future if we want (a client can always escape with a backslash, we forgot that in the spec).

**Comment by felixfbecker** [comment]
Maybe we can get some Eclipse Che folks to comment on this?

**Comment by sqs** [comment]
@felixfbecker:

> Hm, hard to judge now for possible later use cases. 

Since globbing is an optimization, what about making the method `workspace/files` and making it take an `interface WorkspaceFilesParams {}` for now and return a list of all filenames. Language servers would be required to post-filter the filenames returned. Then we can add a `patterns?: string[];` field in a future extension, which would be an optimization, when we know more about the various options and syntaxes that would be required across various language servers.

**Comment by felixfbecker** [comment]
Sounds good! Would also make it clear that no directories are returned, only files.

Another alternative to glob support would be a regexp-based filter, which would only be a few lines of code in any language. `ClientCapabilities` could be extended in a way that let's the client express if he supports regexp filter, glob, both. But if we want to keep it simple for now let's just return a plain files list, and jump straight to globbing later if it really impacts perf (because that is nicer for exposing to the user than regexp).

**Comment by felixfbecker** [comment]
What about the use case of getting files outside the workspace though, like `/usr/local/go`?

**Comment by sqs** [comment]
> What about the use case of getting files outside the workspace though, like /usr/local/go?

Good point. And this would be necessary for globs as well, since it'd be non-ideal to have to specify `/usr/local/go` as `../../../../usr/local/go` (if globs were interpreted as relative to the rootPath).

So, `WorkspaceFilesParams` can have a `base: string;` field, which is the URI underneath which files are listed. In the future (as described above) we can add a `patterns?: string[];` field, and those patterns will be relative to this `base` value.

**Comment by felixfbecker** [comment]
Yes, nice and forward-compatible.
The server will then only be able to use `file://` URIs to access stuff outside of the workspace (because he has no way to find out the protocol, host and port as `rootPath` is not a URI), while the whole workspace can operate through any other protocol like `ftp://`, on a remote host. If that is okay for the LS that need this I think this should be fine (would be solved by https://github.com/Microsoft/language-server-protocol/issues/113)

File: extension-glob-content.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+Note that unlike other requests in LSP, these requests are sent by the language server to the client, not vice versa. The client, not the language server, is assumed to have full access to the workspace's contents.
+
+The language server is allowed to request file paths outside of the workspace root. (This is common for, e.g., system dependencies.) The client may choose whether or not to satisfy these requests.
+
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ClientCapabilities {
```
Comment (emidoots) [inline extension-glob-content.md]: This is describing a change to `ClientCapabilities` but your change to `protocol.md` below is to `ServerCapabilities` -- only one is correct, right?

File: protocol.md [hunk 5]
```diff
@@ -1209,6 +1214,18 @@ _Response_:
...
+
+The get symbol request is sent from the client to the server to resolve the definition location and symbol information for a given text document position.
+
+_Request_
+* method: 'textDocument/symbol'
+* params: [`TextDocumentPositionParams`](#textdocumentpositionparams)
+
+_Response_:
+* result: `SymbolInformation` | `SymbolInformation`[]
+* error: code and message set in case an exception happens during the definition request.
```
Comment (emidoots) [inline protocol.md]: Actually, I'm even a little more confused. This change looks completely unrelated to the PR description + `extension-glob-content.md`, hmm

**Comment by sqs** [comment]
Yeah I'm not sure how those changes got in there. Will look soon.

**Comment by felixfbecker** [comment]
@sqs They are included as the first commit because at the time it was a commit on master. Now that @uforic moved it into a branch and updated master with upstream it is included. I will rebase to remove it

File: extension-files.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+Absolute:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "workspace/glob",
+  "params": {
+    "base": ["/usr/local/go/"]
```
Comment (sqs) [inline extension-files.md]: Should be a URI, and a string not a `string[]`

File: extension-files.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+}
+```
+
+Absolute:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "id": 1,
+  "method": "workspace/glob",
```
Comment (sqs) [inline extension-files.md]: workspace/files
